test/runtime/mmap.c

```diff
@@ -0,0 +1,533 @@
+	    exit(EXIT_FAILURE);
+	}
+
+	/* first attempt should fail */
+	ret = mincore(addr, PAGESIZE, vec);
+	if (!ret) {
+	    fprintf(stderr, "mincore succeeded on untouched mmap'd page\n");
+	    exit(EXIT_FAILURE);
+	}
+
+	memset(addr, 0, PAGESIZE);
+
+	/* must succeed now */
+	__mincore(addr, PAGESIZE, vec);
+	ret = mincore(addr, PAGESIZE, vec);
+
+	__munmap(addr, PAGESIZE);
+	
+	/* must fail again */
+	ret = mincore(addr, PAGESIZE, vec);
+	if (!ret) {
+	    fprintf(stderr, "mincore succeeded on unmapped page\n");
+	    exit(EXIT_FAILURE);
+	}
+    }
+
+    printf("** all mincore tests passed\n"); 
+}
+
+/*
+ * mremap tests
+ *
+ * we test two different behaviors:
+ *  (i)  shrinking mmap
+ *  (ii) expanding mmap
+ */
+#define MREMAP_INIT_SIZE	(1ULL << 12)
+#define MREMAP_END_SIZE		(1ULL << 31)
+#define MREMAP_MOVE_INC		(1ULL << 21)
+#define MREMAP_NR_INCS		(1ULL << 10)
+#define MREMAP_NR_MMAPS		(1ULL << 9)
+
+void mremap_test(void)
+{
+    void * map_addr, * tmp;
+    unsigned long map_size;
+    mmap_t * mmaps;
+    int i;
+
+    printf("** starting mremap tests\n");
+
+    map_addr = mmap(NULL, MREMAP_INIT_SIZE, PROT_READ|PROT_WRITE,
+	    MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (map_addr == MAP_FAILED) {
+	perror("mmap failed");
+	exit(EXIT_FAILURE);
+    }
+
+    /* ensure that MREMAP_MAYMOVE is needed and MREMAP_FIXED is invalid */
+    {
+	tmp = mremap(map_addr, MREMAP_INIT_SIZE, MREMAP_INIT_SIZE*2, 0);
+	if (tmp != MAP_FAILED) {
+	    fprintf(stderr, "mremap succeeded without MREMAP_MAYMOVE??\n");
+	    exit(EXIT_FAILURE);
+	}
+
+	tmp = mremap(map_addr, MREMAP_INIT_SIZE, MREMAP_INIT_SIZE*2, 
+		MREMAP_MAYMOVE | MREMAP_FIXED, map_addr+PAGESIZE);
+	if (tmp != MAP_FAILED) {
+	    fprintf(stderr, "mremap succeeded with MREMAP_FIXED??\n");
+	    exit(EXIT_FAILURE);
+	}
+    }
+
+    /*
+     * allocate a bunch of mmaps to create a fragmentes address space
+     */
+    mmaps = malloc(sizeof(mmap_t) * MREMAP_NR_MMAPS);
+    if (mmaps == NULL) {
+	perror("mmap failed");
+	exit(EXIT_FAILURE);
+    }
+
+    for (i = 0; i < MREMAP_NR_MMAPS; i++) {
+	unsigned long size = gen_random_size();
+	tmp = mmap(
+	    NULL, 
+	    size, 
+	    PROT_READ | PROT_WRITE,
+	    MAP_PRIVATE | MAP_ANONYMOUS, 
+	    -1, 0
+	);
+
+	if (tmp == MAP_FAILED) {
+	    perror("mmap failed");
+	    exit(EXIT_FAILURE);
+	}
+
+	mmaps[i].addr = tmp;
+	mmaps[i].size = size;
+    }
+
+    /* now, remap the inital mmap a bunch of times */ 
+    map_size = MREMAP_INIT_SIZE;
+    for (i = 0; i < MREMAP_NR_INCS; i++) {
+	/* round new_size to the next largest MREMAP_MOVE_INC boundary */
+	unsigned long new_size = 
+		(i == 0) ? MREMAP_MOVE_INC : map_size+MREMAP_MOVE_INC;
+
+	tmp = mremap(map_addr, map_size, new_size, MREMAP_MAYMOVE);
+	if (tmp == MAP_FAILED) {
+	    perror("mremap failed");
+	    exit(EXIT_FAILURE);
+	}
+
+	map_addr = tmp;
+	map_size = new_size;
+    }
+
+    printf("** all mremap tests passed\n");
+}
+
+int main(int argc, char * argv[])
+{
+    mmap_test();
```

Comment:
You may want to unbuffer stdout here, e.g. "setbuf(stdout, NULL)" so that printfs are in line with any possible error messages from nanos. (And then the fflushes above can be removed.)

Reply:
thanks for the tip